src/Saturn/ControllerHelpers.fs

```diff
@@ -40,6 +42,47 @@ module ControllerHelpers =
     let file (ctx: HttpContext) path =
       ctx.WriteHtmlFileAsync path
 
+    ///Returns to the client response according to accepted content type (`Accept` header, and if it's not present `Content-Type` header)
+    let response (ctx: HttpContext) (output: 'a) =
+      let jsonMediaType = MediaTypeHeaderValue.Parse (StringSegment "application/json")
+      let xmlMediaType = MediaTypeHeaderValue.Parse (StringSegment "application/xml")
+      let plainMediaType = MediaTypeHeaderValue.Parse (StringSegment "text/plain")
+      let htmlMediaType = MediaTypeHeaderValue.Parse (StringSegment "text/html")
+
+      match ctx.Request.Headers.TryGetValue "Accept" with
+      | true, header ->
+        let mediaTypes = MediaTypeHeaderValue.ParseList header
+
+        match typeof<'a> with
+        | k when k = typeof<string> && mediaTypes |> Seq.exists (plainMediaType.IsSubsetOf) -> ctx.WriteTextAsync(unbox<string> output)
+        | k when k = typeof<string> && mediaTypes |> Seq.exists (htmlMediaType.IsSubsetOf) -> ctx.WriteHtmlStringAsync(unbox<string> output)
```

Comment:
This approach doesn't do anything about the "quality" field of Accept headers (see https://tools.ietf.org/html/rfc7231#page-38 for details). If the client has specified an Accept header that says "I have a 0.5 quality preference for `text/plain` but a 1.0 quality preference for `text/html`", RFC 7231 says that the server SHOULD respond with a `text/html` content type since that's the user's expressed preference. It might be better to change the logic here to first sort the accepted media types in descending quality order, then run a `Seq.tryFind` where the predicate looks at each media type in turn and sees if `'a` can be rendered as that media type.

The other thing that may be a problem is preferring text/plain for string values. What if the controller's `indexAction` (among others) calls an HTML-rendering function from some other view engine (Razor, Liquid, etc) and returns its result as a string? E.g.,

```fsharp
controller {
    index (fun _ -> task {
        return "<html><body>etc.</body></html>"
    })
    add (fun _ -> task {
        return someHtmlRenderingFunction()
    })
}
```

This one is a bit tricky, because returning a string from a controller might be intended as plain text or as HTML. GiraffeViewEngine obviously doesn't have this issue because its return type is XmlNode, but I feel like this logic will end up bothering people who use other view engines.

OTOH, if people want their strings to be seen as HTML, there's always the `render` function in ControllerHelpers, so maybe `text/plain` should be the default for a plain string returned in the Controller after all. I'll have to think about this one.

Reply:
> This approach doesn't do anything about the "quality" field of Accept headers

TBF, I don't think I ever seen anyone using `quality` field ;-)

> The other thing that may be a problem is preferring text/plain for string values

We can change the order, and prefer `text/html`, I don't see any problems with it. In the end we need to choose some convention - people have ability to fallback to lower level (i.e returning standard HttpFuncResult) if they need more control.

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation/List_of_default_Accept_values suggests that all the major non-MS browsers (Firefox, Chrome, Safari) send a `*/*` request with `q=0.8`. (The `Accept` header is confusing because `;` is the "inner" separator while `,` is the "outer" one, so `text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8` means that you want either `text/html` or `application/xhtml+xml` at the default value of 1.0, `application/xml` at the value of 0.9, and `*/*` at the value of 0.8). https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.formatters.mediatype.issubsetof?view=aspnetcore-2.1 doesn't tell me the logic involved in `IsSubsetOf`: would `text/plain` be considered a subset of `*/*`, or not? If `IsSubsetOf` would consider `text/plain` to be a subset of `*/*`, then the current logic would hand `text/plain` instead of `text/html` to a default-settings Firefox browser, the opposite of what the RFC says it should do.

I think that defaulting to `text/html` is the more practical thing to do.

Reply:
Here is link to `IsSubsetOf` doc - https://docs.microsoft.com/en-us/dotnet/api/microsoft.net.http.headers.mediatypeheadervalue.issubsetof?view=aspnetcore-2.1#Microsoft_Net_Http_Headers_MediaTypeHeaderValue_IsSubsetOf_Microsoft_Net_Http_Headers_MediaTypeHeaderValue_